core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala

```diff
@@ -761,7 +786,10 @@ class TimeSeriesShard(val dataset: Dataset,
        These keys are from earliest time bucket that are still ingesting */
       val earliestTimeBucket = cmd.timeBucket - numTimeBucketsToRetain
       if (earliestTimeBucket >= 0) {
-        val partIdsToRollOver = timeBucketBitmaps.get(earliestTimeBucket).and(activelyIngesting)
+        val bitmap = timeBucketBitmaps.get(earliestTimeBucket)
+        val stamp = activelyIngestingLock.readLock()
```

Comment:
The read lock should be obtained before the `timeBucketBitmaps.get()`, which is the actual read.  The and operation below generates a new bitmap, it doesn't modify anything in place.  Furthermore, the read lock should cover down to line 793 where we do the `timeBucketBitmaps.get(cmd.timeBucket)` as that is also a read operation.

Reply:
The call to timeBucketBitmaps.get() retrieves a pointer from a thread-safe map. No additional locking is required. The 'and' operation needs a read lock to ensure stability of the activelyIngesting bitmap being accessed. Line 793 is only accessing the thread-safe map and the locally accessible bitmap.

Reply:
The timeBucketBitmaps may be a thread safe map but the order of operations here is not atomic.  Couldn't you have a situation like this:

Flush task 1: timeBucketBitmaps.get()
Flush task 2: timeBucketBitmaps.get()
Flush task 1: timeBucketBitmaps.put() with modified map
Flush task 2: timeBucketBitmaps.put(). <-- here overwrites map modified by flush task 1

Reply:
If this pattern is possible, then yes, all bitmaps must be protected.

Reply:
@broneill  @velvia Ok - I now recall why this code is ok. Every flush task reads bits on the earliest time bucket bitmap and sets bits on the latest timeBucket, both of which are uniquely associated with the flush group. Since each flush group is associated with different latest and earliest time buckets, concurrent threads should not be reading or writing to same time bucket bitmaps, or even setting the same time bucket in the collection. This can in theory happen only if a flush task lasts more than the retention period (not possible).

I wish I had documented this clearly earlier in the code. 

@broneill if you could add these comments to the relevant lines, it would really help.